src/s2geography/geoarrow_test.cc

```diff
@@ -202,3 +208,149 @@ TEST(GeoArrow, GeoArrowReaderReadWKTCollection) {
+
+  s2geography::geoarrow::ExportOptions options;
+  options.set_projection(s2geography::pseudo_mercator());
+  InitSchemaGeoArrowPoint(schema.get());
+  writer.Init(schema.get(), options);
+
+  writer.WriteGeography(*geog1);
+  writer.WriteGeography(*geog2);
+  writer.Finish(array.get());
+
+  EXPECT_EQ(array->length, 2);
+
+  auto xs = reinterpret_cast<const double*>(array->children[0]->buffers[1]);
+  EXPECT_DOUBLE_EQ(xs[0], 0.0);
+  EXPECT_DOUBLE_EQ(xs[1], 222638.98158654661);
+  auto ys = reinterpret_cast<const double*>(array->children[1]->buffers[1]);
+  EXPECT_DOUBLE_EQ(ys[0], 111325.14286638441);
+  EXPECT_DOUBLE_EQ(ys[1], 334111.17140195851);
+}
+
+TEST(GeoArrow, GeoArrowWriterPolylineTessellated) {
+  s2geography::WKTReader reader;
+  auto geog = reader.read_feature("LINESTRING (-64 45, 0 45)");
+
+  Writer writer;
+  nanoarrow::UniqueSchema schema;
+  GeoArrowSchemaInitExtension(schema.get(), GEOARROW_TYPE_LINESTRING);
+
+  nanoarrow::UniqueArray array;
+  writer.Init(schema.get());
+  writer.WriteGeography(*geog);
+  writer.Finish(array.get());
+
+  EXPECT_EQ(array->length, 1);
+  EXPECT_EQ(array->children[0]->length, 2);
+
+  // with tessellation -> more coordinates
+  nanoarrow::UniqueArray array2;
+  s2geography::geoarrow::ExportOptions options;
+  options.set_tessellate_tolerance(S1Angle::Radians(10000 / EARTH_RADIUS_METERS));
+  writer.Init(schema.get(), options);
+  writer.WriteGeography(*geog);
+  writer.Finish(array2.get());
+
+  EXPECT_EQ(array2->length, 1);
+  EXPECT_GT(array2->children[0]->length, 2);
+  EXPECT_EQ(array2->children[0]->length, 9);
+  // first coordinate is still the same
+  auto xs = reinterpret_cast<const double*>(array2->children[0]->children[0]->buffers[1]);
+  EXPECT_DOUBLE_EQ(xs[0], -64);
+  EXPECT_DOUBLE_EQ(xs[9], 0);
+  auto ys = reinterpret_cast<const double*>(array2->children[0]->children[1]->buffers[1]);
+  EXPECT_DOUBLE_EQ(ys[0], 45);
+  EXPECT_DOUBLE_EQ(ys[8], 45);
+  EXPECT_GT(ys[4], 45);
+
+  // with tessellation and projection
+  nanoarrow::UniqueArray array3;
+  options.set_projection(s2geography::pseudo_mercator());
+  writer.Init(schema.get(), options);
+  writer.WriteGeography(*geog);
+  writer.Finish(array3.get());
+
+  EXPECT_EQ(array3->length, 1);
+  EXPECT_GT(array3->children[0]->length, 2);
+  EXPECT_EQ(array3->children[0]->length, 9);
+
+  xs = reinterpret_cast<const double*>(array3->children[0]->children[0]->buffers[1]);
+  EXPECT_NEAR(xs[0], -7124447.41, 0.01);
+  EXPECT_DOUBLE_EQ(xs[9], 0);
+  ys = reinterpret_cast<const double*>(array3->children[0]->children[1]->buffers[1]);
+  EXPECT_NEAR(ys[0], 5621521.48, 0.01);
+  EXPECT_NEAR(ys[8], 5621521.48, 0.01);
+  EXPECT_GT(ys[4], 5621521);
+}
+
+TEST(GeoArrow, GeoArrowWriterPolygonTessellated) {
+  s2geography::WKTReader reader;
+  auto geog = reader.read_feature("POLYGON ((-64 45, 0 45, 0 55, -64 55, -64 45))");
+
+  Writer writer;
+  nanoarrow::UniqueSchema schema;
+  GeoArrowSchemaInitExtension(schema.get(), GEOARROW_TYPE_WKT);
```

Comment:
I was first planning to do this with `GEOARROW_TYPE_POLYGON` like I did for lines above, but, this seems to be buggy, as it gives an empty arrow array as result (I saw this both in the tests here as trying the same through `spherely`). Given that the output to WKT does work correctly, my first assumption is that this is rather a bug in geoarrow-c than in the code here in s2geography.

Reply:
Ah, no, it _was_ a bug on the s2geography side, there was a `feat_end` visitor call missing, and WKT or point/linestring output doesn't seem to care about that, but for polygon that fixed the issue. Good that I was testing that! ;)

Reply:
So fixed that and added a bunch of roundtrip tests as well (those are now covering more types, but with the default of no tessellation)